### PR TITLE
[Performance] RGDR optimization

### PR DESCRIPTION
**Pull request by vladimirevmenoff**
### What this PR does / why we need it?
This PR optimizes the RecurrentGatedDeltaRule (RGDR) operator for Ascend 310P. Key optimizations include:
- Batched `WholeReduceSum` to reduce instruction dispatch overhead.
- Out-of-place reduction fold to eliminate UB bank group conflicts.
- Phased barrier restructuring for better pipeline efficiency.
- Simplified `MatVecMul` loop structure.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Verified with cosine similarity >= 0.99 against CPU golden on simulator.
- Performance profiling on 310P3 showing ~52% reduction in scalar pipe cycles and ~88% reduction in bank group conflicts.

Screenshot 2026-05-25 at 11 16 18

Screenshot 2026-05-25 at 11 25 58

- vLLM version: v0.20.2
- vLLM main: https://github.com/vllm-project/vllm/commit/39910f2b25aacc09f5e7f166cdf0030b19f8b9e8

### CODE REVIEW

File: csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h [hunk 1]
```diff
@@ -329,115 +348,118 @@ class RGDR {
...
+        for (uint32_t j = 0; j < rows; j += MAX_REPEAT_TIME) {
+            uint32_t batch = Std::min(static_cast<uint32_t>(MAX_REPEAT_TIME), rows - j);
+            if (readFromSrc) {
+                WholeReduceSum(dstTensor[j], srcTensor[j * alignK_],
+                               REPEAT_LENTH, batch, 1, 1, foldStride);
+            } else {
+                WholeReduceSum(dstTensor[j], foldTmpUb[j * curK],
+                               REPEAT_LENTH, batch, 1, 1, foldStride);
+            }
         }
```
[the remark this answers is not in the document]
Comment (vladimirevmenoff) [reply csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h:460]: This claim is wrong. It misunderstands the dstRepStride unit.

  WholeReduceSum's dstRepStride is in sizeof(T) bytes, NOT DataBlocks (32B). This
   is documented and differs from other V-pipe instructions (Mul, Add, etc.)
  which use 32B blocks.

  For float: dstRepStride=1 = 4 bytes = 1 element apart = consecutive floats.
  Results land at dst[0], dst[1], dst[2], ... — no gap, no overflow.
  
  Our call:
  WholeReduceSum(dstTensor[j], srcTensor[j*alignK_], REPEAT_LENTH, batch, 1, 1,
  foldStride);
 
  batch consecutive floats written starting at dstTensor[j]. batch <= vStep_ =
  buffer size. No overflow.
  
  Verified empirically — test_v310.py passes all shapes with max_err < 5e-4
  against fp32 golden. If results were 8 floats apart, we'd get garbage, not 1e-4
   accuracy.

File: csrc/attention/recurrent_gated_delta_rule_v310/examples/python/compare_vllm.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
+"""Compare ops-transformer RGDR (ctypes) vs vLLM RGDR (torch.ops._C_ascend) vs golden CPU ref.
+
+Usage:
+    # on 310P device with both ops-transformer and vllm-ascend installed:
+    python3 compare_vllm.py [--dtype fp16|fp32] [--qwen]
+"""
+
+import os
```
Comment (Tflowers-0129) [inline csrc/attention/recurrent_gated_delta_rule_v310/examples/python/compare_vllm.py]: The placement of these files in examples/python still needs to be considered, and they need to be added to the nightly monitoring.

File: csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
+ * CANN Open Software License Agreement Version 2.0 (the "License").
  * Please refer to the License for details. You may not use this file except in compliance with the License.
  * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
  * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  * See LICENSE in the root of the software repository for the full text of the License.
  */
 
 /*!
- * \file recurrent_gated_delta_rule_v310.h
+ * \file grouped_matmul_finalize_routing.h
```
Comment (Tflowers-0129) [inline csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h]: why grouped_matmul_finalize_routing？

File: vllm_ascend/_310p/model_runner_310p.py [hunk 4]
```diff
@@ -222,7 +222,7 @@ def _prepare_inputs(  # type: ignore[override]
             out=positions_np,
         )
         block_table = cast(MultiGroupBlockTable310, self.input_batch.block_table)
-        block_table.compute_slot_mapping(
+        block_table.compute_slot_mapping_draft(
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/model_runner_310p.py]: no need any more, the CI now is OK

File: vllm_ascend/_310p/block_table.py [hunk 5]
```diff
@@ -14,9 +14,12 @@
...
     def compute_slot_mapping(self, *args: Any) -> None:
         req_indices, positions = self._normalize_slot_mapping_inputs(*args)
-        self._compute_slot_mapping_numpy(req_indices, positions)
+        self._compute_slot_mapping_np(req_indices, positions)
 
-    def _compute_slot_mapping_numpy(self, req_indices: np.ndarray, positions: np.ndarray) -> None:
+    def compute_slot_mapping_draft(self, req_indices: np.ndarray, positions: np.ndarray) -> None:
+        self._compute_slot_mapping_np(self._to_numpy(req_indices), self._to_numpy(positions))
+
+    def _compute_slot_mapping_np(self, req_indices: np.ndarray, positions: np.ndarray) -> None:
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/block_table.py]: no need any more, the CI now is OK

File: csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h [hunk 6]
```diff
@@ -542,10 +581,9 @@ class RGDR {
                 nextStateOffset = ((stateOffset * NV_ + head_i) * realV_ + nextVOffset) * realK_;
                 PrefetchState(nextStateOffset, nextSingleV);
             }
-            uint64_t pendingAttnOffset = 0;
-            uint64_t pendingStateOffset = 0;
-            bool hasPendingAttn = false;
-            bool hasPendingState = false;
+            evtMte3V_ = static_cast<event_t>(GetTPipePtr()->FetchEventID(HardEvent::MTE3_V));
+            evtVMte3_ = static_cast<event_t>(GetTPipePtr()->FetchEventID(HardEvent::V_MTE3));
+            SetFlag<HardEvent::MTE3_V>(evtMte3V_);
```
Comment (Tflowers-0129) [inline csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h:585]: Does it need to be released in pairs to ensure there is no leak?

Comment (vladimirevmenoff) [reply csrc/attention/recurrent_gated_delta_rule_v310/op_kernel/recurrent_gated_delta_rule_v310.h:585]: Of course, otherwise its a dead lock

The code has all matching Set/Wait flags right now.

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_recurrent_gated_delta_rule_v310.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import os
+
+import pytest
+import torch
+import torch_npu
+
+torch_npu.npu.set_compile_mode(jit_compile=False)
+
+_CANN = os.environ.get("ASCEND_HOME_PATH", "/usr/local/Ascend/ascend-toolkit/latest")
+_CUST = f"{_CANN}/opp/vendors/custom_transformer/op_api/lib"
```
Comment (Tflowers-0129) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_recurrent_gated_delta_rule_v310.py]: remove this ut pls, already in other file 'tests/e2e/nightly/310p/....'

**Review by ZT-AIA — APPROVED** [review]

**Comment by Tflowers-0129** [comment]
RFC : #9711
